File: fsharp/binary-search/BinarySearch.fs
module BinarySearch

let rec find (input: int array) (value: int): int option = 
    if input.Length = 0 then
        None
    else
        let middle = input.Length /2
        if input.[middle] = value then 
            Some middle
        elif input.[middle] > value then
            find input.[..(middle-1)] value
        else 
            match find input.[(middle+1)..] value with
            | None -> None
            | Some i -> Some (middle + 1 + i)

File: csharp/binary-search/BinarySearch.cs
using System;

public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        if (input.Length == 0) return -1;
        int middle = input.Length / 2;
        if (input[middle] == value) 
            return middle;
        else if (input[middle] > value) 
            return Find(input[0..middle], value);
        else 
            return Find(input[(middle + 1)..], value) == -1 ? -1 : middle + 1 + Find(input[(middle + 1)..], value);
    }
}